pkg/operator/latencyprofilecontroller/controller.go

```diff
@@ -0,0 +1,195 @@
+		"WorkerLatencyProfile",
+		eventRecorder.WithComponentSuffix("latency-profile-controller"),
+	)
+}
+
+func (c *LatencyProfileController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+
+	// Collect the current latency profile
+	configNodeObj, err := c.configNodeLister.Get("cluster")
+
+	// if config/v1/node/cluster object is not found this controller should do nothing
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.Warningf("nodes.config.openshift.io/cluster not found")
+			return nil
+		}
+		return err
+	}
+
+	// in case of empty workerlatency profile, set status Complete=False and pre-empt sync
+	if configNodeObj.Spec.WorkerLatencyProfile == "" {
+		// TODO: in case worker latency profile is transitioned
+		// from Default/Medium/Low to ""
+		// check that the required args should vanish
+		_, err = c.updateStatus(ctx, reasonLatencyProfileEmpty, "latency profile not set on cluster", false, true)
+		return err
+	}
+
+	_, operatorStatus, _, err := c.operatorClient.GetStaticPodOperatorState()
+	if err != nil {
+		return err
+	}
+
+	// Collect the unique set of revisions for all nodes in cluster
+	revisionMap := map[int32]struct{}{}
+	uniqueRevisions := []int32{}
+	for _, nodeStatus := range operatorStatus.NodeStatuses {
+		revision := nodeStatus.CurrentRevision
+		if _, ok := revisionMap[revision]; !ok {
+			revisionMap[revision] = struct{}{}
+			uniqueRevisions = append(uniqueRevisions, revision)
+		}
+	}
+
+	// Collect the current latency profile and desired arg val pairs
+	desiredArgValPairs, ok := c.latencyProfileArgValMap[configNodeObj.Spec.WorkerLatencyProfile]
+	// In case of unknown latency profile
+	if !ok {
+		err := fmt.Errorf("unknown worker latency profile found in nodes.config.openshift.io/cluster Spec.WorkerLatencyProfile = %v", configNodeObj.Spec.WorkerLatencyProfile)
+		return err
+	}
+
+	// For each revision, check that the configmap for that revision have correct arg val pairs or not
+	revisionsHaveSynced := true
+	for _, revision := range uniqueRevisions {
+		configMapNameWithRevision := fmt.Sprintf("%s-%d", revisionConfigMapName, revision)
+		configMap, err := c.configMapClient.ConfigMaps(c.targetNamespace).Get(ctx, configMapNameWithRevision, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		match, err := configMatchProfileArguments(configMap, desiredArgValPairs)
+		if err != nil {
+			return err
+		}
+		if !match {
+			revisionsHaveSynced = false
+			break
+		}
+	}
+
+	if revisionsHaveSynced {
+		_, err = c.updateStatus(ctx, reasonLatencyProfileUpdated, "all static pod revision(s) have updated latency profile", false, true)
+	} else {
+		_, err = c.updateStatus(ctx, reasonLatencyProfileUpdateTriggered, "one or more static pod revision(s) are updating latency profile", true, false)
+	}
+	return err
+}
+
+func configMatchProfileArguments(configMap *corev1.ConfigMap, desiredArgValPairs []nodeobserver.LatencyProfileArgVal) (bool, error) {
```

Comment:
I think this function should be provided by the operator.

Reply:
We could always keep such a function in the operator but what I found was that it would lead to code repetition in cluster-kube-apiserver-operator and cluster-kube-controller-manager-operator codebase atleast. Not sure if wanna repeat the same logic in both of the operators.

Moreover, I see that this function is kinda generic as the intended behavior for this function is to compare content present in installercontroller generated config maps (storing configs about different active revisions) against known values.

Reply:
(please nvm, this is slightly contd. from https://github.com/openshift/library-go/pull/1364#discussion_r867704531)

Decoupled the logic for checking revision matching from the actual controller (as it should be operator specific), moved this functionality over to `pkg/operator/latencyprofilecontroller/helpers` by crafting a more generic implementation that returns a `MatchProfileRevisionConfigsFunc` which can be passed onto latencyprofilecontroller at init time.

Essentially, would like to invoke `helpers.NewInstallerRevisionConfigMatcher(...)` in the operator(s) (KAO, KCMO in our case), the output from which could be passed to our controller.